### PR TITLE
Validator

### PR DESCRIPTION
**Pull request by GabFitzgerald**
Added a ULID validator to check that a string is a valid ULID. 
Checks the string is the correct length and only contains valid characters.

Also started a seed_time decoder, would appreciate the feedback.

### CODE REVIEW

File: spec/ulid_spec.cr [hunk 1]
```diff
@@ -56,5 +56,36 @@ describe ULID do
         (ulid_2 < ulid_1).should be_true
       end
     end
+
+    it "validator should return true if a string is a valid ulid" do
+      validator = ULID.valid?("01B3EAF48P97R8MP9WS6MHDTZ3")
+      validator.should eq true
+    end
+
+    it "should not valdate invalid strings" do
```
Comment (kimburgess) [inline spec/ulid_spec.cr]: Pop and empty string check in here too - good base / edge case to provide test coverage for.

File: spec/ulid_spec.cr [hunk 2]
```diff
@@ -56,5 +56,36 @@ describe ULID do
...
+      ULID.valid?("abcde").should eq false
+      ULID.valid?("1234567890").should eq false
+      ULID.valid?("01!3EAF48P97R8MP9WS8MHDTZ3").should eq false
+    end
+
+    # test seed_time method
+    it "should correctly decode a ulid's seed time" do
+      seedtime = ULID.seed_time("01B3EAF48P97R8MP9WS6MHDTZ3")
+      seedtime.should be_a Time
+      # seedtime.should eq (2016-12-08 04:18:39.001000000 UTC)
```
Comment (kimburgess) [inline spec/ulid_spec.cr]: Worth re-enabling these.

File: src/ulid.cr [hunk 3]
```diff
@@ -19,6 +19,23 @@ module ULID
...
+    # is ulid the correct size AND only contain valid characters?
+    ulid.size == TIME_LEN + RANDOM_LEN && (ulid =~ /[^0123456789ABCDEFGHJKMNPQRSTVWXYZ]/).nil?
+  end
+
+  # TODO #valid!(ulid : String) => returns nil, raise appropriate errors in invalid
+
+  def seed_time(ulid : String) : Time
+    Time.unix_ms(ulid[0..9].to_i64(32))
+  end
+
```
Comment (kimburgess) [inline src/ulid.cr:64]: A nice addition to make you be `ULID.parse(String) : { Time, String }`. This can check the format is valid (raising an exception if not), the parse out the time and random components.

File: src/ulid.cr [hunk 4]
```diff
@@ -19,6 +19,23 @@ module ULID
...
 
+  # Validate a string is a ULID
+  #
+  # ```
+  # ULID.valid?("01B3EAF48P97R8MP9WS6MHDTZ3")
+  # # => true
+  # ```
+  def valid?(ulid : String) : Bool
+    # is ulid the correct size AND only contain valid characters?
+    ulid.size == TIME_LEN + RANDOM_LEN && (ulid =~ /[^0123456789ABCDEFGHJKMNPQRSTVWXYZ]/).nil?
```
Comment (kimburgess) [inline src/ulid.cr]:
```suggestion
    # Incorrect length
    return false unless ulid.size == TIME_LEN + RANDOM_LEN

    # Invalid chars
    return false unless ulid.chars.all? &.in?(ENCODING)

    return true
```

File: src/ulid.cr [hunk 5]
```diff
@@ -19,6 +19,57 @@ module ULID
...
+
+  # Validate a string is a ULID, raises errors for invalid strings
+  #
+  # ```
+  # ULID.valid!("01B3EAF48P97R8MP9WS6MHDTZ3")
+  # # => nil
+  # ```
+   def valid!(ulid : String) : Nil
+    raise IncorrectLength.new("ULID should be 26 characters") unless ulid.size == TIME_LEN + RANDOM_LEN
+    raise InvalidChars.new("Invalid characters found in string. Should only contain #{ENCODING}") unless ulid.upcase.chars.all? &.in?(ENCODING)
```
Comment (GabFitzgerald) [inline src/ulid.cr]: Should 
`ulid.size == TIME_LEN + RANDOM_LEN `
and 
`ulid.upcase.chars.all? &.in?(ENCODING) `
be refactored into their own methods?

File: src/ulid.cr (hunk 3 above)
Comment (GabFitzgerald) [reply src/ulid.cr:64]: Started an initial attempt. Have moved that work in progress over to another branch to PR later.

File: src/ulid.cr (hunk 5 above)
Comment (kimburgess) [reply src/ulid.cr]: I'd keep them together. There's no additional information outside of if it's a valid ULID or not.

On a similar note, I would avoid raising exceptions here. As a user of the lib, the main interest is just checking validity, which can be handled most simply with a plain, boring `ULID.valid?(String) : Bool`. There is no possible exceptional circumstance that can occur based on what is is encoded into the type `String -> Bool`.

File: spec/ulid_spec.cr [hunk 6]
```diff
@@ -56,5 +56,39 @@ describe ULID do
         (ulid_2 < ulid_1).should be_true
       end
     end
+
+    describe ".valid? : Bool" do
+      it "validate a valid string" do
+        ULID.valid?("01B3EAF48P97R8MP9WS6MHDTZ3").should eq true
+      end
+
+      it "should not validate invalid strings" do
```
Comment (kimburgess) [inline spec/ulid_spec.cr]: Recommend splitting this up into what is being tested. E.g. `it "detects incorrect length"`, `it "detects invalid characters"` etc

File: spec/ulid_spec.cr [hunk 7]
```diff
@@ -56,5 +56,39 @@ describe ULID do
         (ulid_2 < ulid_1).should be_true
       end
     end
+
+    describe ".valid? : Bool" do
+      it "validate a valid string" do
+        ULID.valid?("01B3EAF48P97R8MP9WS6MHDTZ3").should eq true
```
Comment (kimburgess) [inline spec/ulid_spec.cr]:
```suggestion
        ULID.valid?("01B3EAF48P97R8MP9WS6MHDTZ3").should be_true
```

File: spec/ulid_spec.cr [hunk 8]
```diff
@@ -56,5 +56,39 @@ describe ULID do
...
+        ULID.valid?("01!3EAF48P97R8MP9WS8MHDTZ3").should eq false
+        ULID.valid?("").should eq false
+      end
+    end
+
+    describe ".seedtime : Time" do
+      it "should correctly decode seed time" do
+        seedtime = ULID.seed_time("01B3EAF48P97R8MP9WS6MHDTZ3")
+        seedtime.should be_a Time
+        seedtime.should eq Time.utc(2016, 12, 8, 4, 18, 39, nanosecond: 1000000)
```
Comment (kimburgess) [inline spec/ulid_spec.cr]: Check out [`Time.unix_ms`](https://crystal-lang.org/api/0.35.1/Time.html#unix_ms(milliseconds:Int):Time-class-method)

File: src/ulid.cr [hunk 9]
```diff
@@ -19,6 +19,35 @@ module ULID
     encode_time(seed_time, TIME_LEN) + encode_random(RANDOM_LEN)
   end
 
+  class IncorrectLength < Exception; end
+
+  class InvalidChars < Exception; end
```
Comment (kimburgess) [inline src/ulid.cr]: These don't looks to be in use.

File: src/ulid.cr [hunk 10]
```diff
@@ -19,6 +19,35 @@ module ULID
...
+  end
+
+  # Decode ULID seedtime
+  #
+  # ```
+  # ULID.seed_time("01EVDK4C0Q275A7AHHEVX02DCG")
+  # # => 2021-02-25 22:14:43.994000000 UTC
+  # ```
+  def seed_time(ulid : String) : Time
+    Time.unix_ms(ulid[0..9].to_i64(32))
```
Comment (kimburgess) [inline src/ulid.cr]: This is incorrect. You'll need to invert the process used in `#encode_time`. Importantly, the alphabet used for encoding is _Crockford 32_ rather than vanilla base32.

Comment (GabFitzgerald) [reply src/ulid.cr]: Ah yep. I'll move this method into a separate PR.

File: src/ulid.cr (hunk 9 above)
Comment (GabFitzgerald) [reply src/ulid.cr]: In valid! method now

File: spec/ulid_spec.cr [hunk 11]
```diff
@@ -56,5 +56,51 @@ describe ULID do
         (ulid_2 < ulid_1).should be_true
       end
     end
+
+    describe ".valid?" do
+      it "should validate a valid string" do
+        ULID.valid?("01B3EAF48P97R8MP9WS6MHDTZ3").should be_true
+        ULID.valid?("01b3EAF48P97R8MP9WS6MHDTZ3").should be_true # ulids are not case sensitive
+      end
```
Comment (caspiano) [inline spec/ulid_spec.cr]: perhaps split this into 2 specs?

**Review by caspiano — COMMENTED** [review]
looking good!

File: src/ulid.cr (hunk 10 above)
Comment (GabFitzgerald) [reply src/ulid.cr]: Added a working version back into the PR

File: src/ulid.cr [hunk 12]
```diff
@@ -57,11 +57,12 @@ module ULID
...
-    sum = 0_i64
-    ulid[0..9].reverse
-      .each_char_with_index do |char, i|
+    sum = ulid[0..9].reverse
+      .each_char
+      .with_index
+      .sum(0_i64) do |char, i|
         ord = ENCODING.index(char) || raise InvalidChars.new("Character: #{char} not part of Crockford's Base32.")
-        sum += ord.to_i64 * (ENCODING_LEN.to_i64 ** i)
+        ord.to_i64 * (ENCODING_LEN.to_i64 ** i)
```
Comment (caspiano) [inline src/ulid.cr]: Is `ENCODING_LEN` used elsewhere? if not, i would make it i64 to prevent the runtime conversion

File: spec/ulid_spec.cr [hunk 13]
```diff
@@ -57,4 +57,69 @@ describe ULID do
...
+        ULID.valid?("0").should be_false
+        ULID.valid?("01B3EAF48P97R8MP9WS6MHDTZ32").should be_false
+        ULID.valid?("01B3EAF48P97R8MP9WS6MHDTZ").should be_false
+        ULID.valid?("!@#$%^&*(").should be_false
+        ULID.valid?("abcde").should be_false
+        ULID.valid?("1234567890").should be_false
+        ULID.valid?("").should be_false
+      end
+
+    it "should dectect invalid characters" do
```
Comment (caspiano) [inline spec/ulid_spec.cr]: typo: dectect

File: spec/ulid_spec.cr [hunk 14]
```diff
@@ -57,4 +57,69 @@ describe ULID do
...
+      expect_raises(IncorrectLength) { ULID.valid!("0") }
+      expect_raises(IncorrectLength) { ULID.valid!("01B3EAF48P97R8MP9WS6MHDTZ32") }
+      expect_raises(IncorrectLength) { ULID.valid!("01B3EAF48P97R8MP9WS6MHDTZ") }
+      expect_raises(IncorrectLength) { ULID.valid!("!@#$%^&*(") }
+      expect_raises(IncorrectLength) { ULID.valid!("abcde") }
+      expect_raises(IncorrectLength) { ULID.valid!("1234567890") }
+      expect_raises(IncorrectLength) { ULID.valid!("") }
+    end
+    
+    it "should dectect invalid characters" do
```
Comment (caspiano) [inline spec/ulid_spec.cr]: typo: dectect

File: spec/ulid_spec.cr [hunk 15]
```diff
@@ -57,4 +57,69 @@ describe ULID do
...
+      ULID.valid?("01!3EAF48P97R8MP9WS8MHDTZ3").should be_false
+      ULID.valid?("01I3EAF48P97R8MP9WS8MHDTZ3").should be_false
+      ULID.valid?("01O3EAF48P97R8MP9WS8MHDTZ3").should be_false
+    end
+  end
+
+  describe ".valid!" do
+    it "should validate a valid string" do
+      ULID.valid!("01B3EAF48P97R8MP9WS6MHDTZ3").should be_nil
+      ULID.valid!("01b3EAF48P97R8MP9WS6MHDTZ3").should be_nil # ulids are not case sensitive
```
Comment (caspiano) [inline spec/ulid_spec.cr]: same as line 66

**Review by caspiano — APPROVED** [review]

File: src/ulid.cr [hunk 16]
```diff
@@ -19,6 +20,53 @@ module ULID
...
+  # ULID.valid?("01B3EAF48P97R8MP9WS6MHDTZ3")
+  # # => true
+  # ```
+  def valid?(ulid : String) : Bool
+    # Incorrect length
+    return false unless ulid.size == TIME_LEN + RANDOM_LEN
+    # Invalid chars
+    return false unless ulid.upcase.chars.all? &.in?(ENCODING)
+
+    return true
```
Comment (dukenguyenxyz) [inline src/ulid.cr]: I would do this if we are keeping the same structure

```suggestion
    # Incorrect length || Invalid chars
    return false unless ulid.size == TIME_LEN + RANDOM_LEN || ulid.upcase.chars.all? &.in?(ENCODING)
    
    true
```

File: src/ulid.cr [hunk 17]
```diff
@@ -19,6 +20,53 @@ module ULID
...
+  # # => nil (or exception)
+  # ```
+  def valid!(ulid : String) : Nil
+    # Incorrect length
+    raise IncorrectLength.new("ULIDs must be 26 characters.") unless ulid.size == TIME_LEN + RANDOM_LEN
+    # Invalid chars
+    raise InvalidChars.new("Invalid characters found.") unless ulid.upcase.chars.all? &.in?(ENCODING)
+
+    return nil
+  end
```
Comment (dukenguyenxyz) [inline src/ulid.cr]: I would do this

```suggestion
  # Validate a string is a ULID
  #
  # ```
  # ULID.valid?("01B3EAF48P97R8MP9WS6MHDTZ3")
  # # => true
  # ```
  def valid?(ulid : String) : Bool
    valid(ulid, raise_error: false)
  end

  # Validate a string is a ULID
  #
  # ```
  # ULID.valid!("01B3EAF48P97R8MP9WS6MHDTZ3")
  # # => nil (or exception)
  # ```
  def valid!(ulid : String) : Nil
    valid(ulid, raise_error: true)
    nil
  end

  private def valid(ulid : String, raise_error : Bool) : Bool
    # Incorrect length
    length_error = !(ulid.size == TIME_LEN + RANDOM_LEN)
    raise IncorrectLength.new("ULIDs must be 26 characters.") if raise_error && length_error
  
    # Invalid chars
    char_error = !(ulid.upcase.chars.all? &.in?(ENCODING))
    raise InvalidChars.new("Invalid characters found.") if raise_error && char_error
  
    length_error && char_error
  end
```

**Review by dukenguyenxyz — CHANGES_REQUESTED** [review]

File: src/ulid.cr (hunk 16 above)
Comment (caspiano) [reply src/ulid.cr]: or even...
```suggestion
    ulid.size == TIME_LEN + RANDOM_LEN && ulid.upcase.chars.all? &.in(ENCODING)
```

File: src/ulid.cr (hunk 17 above)
Comment (dukenguyenxyz) [reply src/ulid.cr]: Or this to save some some computation but looks a bit less elegant
```crystal
private def valid(ulid : String, raise_error : Bool) : Bool
  # Incorrect length
  unless ulid.size == TIME_LEN + RANDOM_LEN
    return false if !raise_error
    raise IncorrectLength.new("ULIDs must be 26 characters.")
  end

  # Invalid chars
  unless ulid.upcase.chars.all? &.in?(ENCODING)
    return false if !raise_error
    raise InvalidChars.new("Invalid characters found.")
  end

  true
end
```

File: src/ulid.cr (hunk 16 above)
Comment (dukenguyenxyz) [reply src/ulid.cr]: hahaha I think I dropped my brain somewhere today

File: src/ulid.cr (hunk 4 above)
Comment (GabFitzgerald) [reply src/ulid.cr]: @dukeraphaelng @caspiano had this originally, is there a preferred approach?

File: src/ulid.cr [hunk 18]
```diff
@@ -1,11 +1,12 @@
 require "./ulid/*"
+require "big"
```
Comment (kimburgess) [inline src/ulid.cr]: Shouldn't need any `BigInts` here.

File: src/ulid.cr [hunk 19]
```diff
@@ -19,6 +20,49 @@ module ULID
...
+    ulid.size == TIME_LEN + RANDOM_LEN && ulid.upcase.chars.all? &.in?(ENCODING)
+  end
+
+  # Validate a string is a ULID
+  #
+  # ```
+  # ULID.valid!("01B3EAF48P97R8MP9WS6MHDTZ3")
+  # # => nil (or exception)
+  # ```
+  def valid!(ulid : String) : Nil
```
Comment (kimburgess) [inline src/ulid.cr]: Entirely optional, but what about `.validate!` for a little clearer API.

File: src/ulid/exceptions.cr [hunk 20]
```diff
@@ -0,0 +1,3 @@
+class IncorrectLength < Exception; end
+
+class InvalidChars < Exception; end
```
Comment (kimburgess) [inline src/ulid/exceptions.cr]: Having these as seperate error types provides little useful information. For a user, the only concern is it's either a valid ULID, or not. The reason for an invalidity can then be contained in the associated exception message.

These are also currently sitting in the top-level namespace. Definitely a good idea to nest any exception(s) that are defined within the `ULID` module.

File: spec/ulid_spec.cr [hunk 21]
```diff
@@ -57,4 +57,72 @@ describe ULID do
...
+    end
+
+    it "should detect invalid characters" do
+      expect_raises(InvalidChars) { ULID.valid!("01!3EAF48P97R8MP9WS8MHDTZ3") }
+      expect_raises(InvalidChars) { ULID.valid!("01I3EAF48P97R8MP9WS8MHDTZ3") }
+      expect_raises(InvalidChars) { ULID.valid!("01O3EAF48P97R8MP9WS8MHDTZ3") }
+    end
+  end
+
+  describe ".seed_time" do
```
Comment (kimburgess) [inline spec/ulid_spec.cr:105]: Would be good to include a round trip test here (e.g. create a ULID that encodes a specific time, then decode this and ensure they match).

**Review by kimburgess — CHANGES_REQUESTED** [review]
Looking nice! A few minor things as noted.

File: src/ulid.cr [hunk 22]
```diff
@@ -19,6 +20,49 @@ module ULID
...
+    # Incorrect length
+    raise IncorrectLength.new("ULIDs must be 26 characters.") unless ulid.size == TIME_LEN + RANDOM_LEN
+    # Invalid chars
+    raise InvalidChars.new("Invalid characters found.") unless ulid.upcase.chars.all? &.in?(ENCODING)
+
+    return nil
```
Comment (dukenguyenxyz) [inline src/ulid.cr]: no need `return` here as it always return the last value

Comment (dukenguyenxyz) [reply src/ulid.cr]: if you install ameba in your bin folder or install ameba extension in vscode, the linter will pick this up for you

**Review by dukenguyenxyz — APPROVED** [review]

File: src/ulid/exceptions.cr (hunk 20 above)
Comment (caspiano) [reply src/ulid/exceptions.cr]: Definitely agree with this one

Comment (GabFitzgerald) [reply src/ulid/exceptions.cr]: So instead of: 
`raise IncorrectLength.new("ULIDs must be 26 characters.") unless ulid.size == TIME_LEN + RANDOM_LEN`
should I
`raise Exception.new("ULIDs must be 26 characters.") unless ulid.size == TIME_LEN + RANDOM_LEN` ?

Or keep custom exceptions just inside ULID module?

Comment (caspiano) [reply src/ulid/exceptions.cr]: I think custom exception inside of the ULID module makes the most sense to me

Comment (caspiano) [reply src/ulid/exceptions.cr]: You could also move the message inside of the error definition

File: src/ulid.cr (hunk 19 above)
Comment (GabFitzgerald) [reply src/ulid.cr]: nice

should I have
`.valid?` and `validate!`
or
`validate?` and `validate!`

Comment (kimburgess) [reply src/ulid.cr]: I'd go with option a - seems to flow nicely.

File: src/ulid/exceptions.cr (hunk 20 above)
Comment (GabFitzgerald) [reply src/ulid/exceptions.cr]: okay, fixed this up

I'm a little unsure about the file structure
I've just put the exception classes in the ULID module under the `.validate!` method. Is there a more appropriate place?

```crystal
 def validate!(ulid : String) : Nil
    # Incorrect length
    raise IncorrectLength.new() unless ulid.size == TIME_LEN + RANDOM_LEN
    # Invalid chars
    raise InvalidChars.new() unless ulid.upcase.chars.all? &.in?(ENCODING)

    nil
  end

  class IncorrectLength < Exception
    def initialize(message = "ULIDs must be 26 characters.")
      super(message)
    end
  end

  class InvalidChars < Exception
    def initialize(message = "Invalid characters found.")
      super(message)
    end
  end
  ```

File: src/ulid.cr [hunk 23]
```diff
@@ -19,6 +19,61 @@ module ULID
...
+  # # => nil (or exception)
+  # ```
+  def validate!(ulid : String) : Nil
+    # Incorrect length
+    raise IncorrectLength.new() unless ulid.size == TIME_LEN + RANDOM_LEN
+    # Invalid chars
+    raise InvalidChars.new() unless ulid.upcase.chars.all? &.in?(ENCODING)
+
+    nil
+  end
```
Comment (dukenguyenxyz) [inline src/ulid.cr]: can just return void here

File: spec/ulid_spec.cr [hunk 24]
```diff
@@ -57,4 +57,82 @@ describe ULID do
...
+    end
+
+    it "should detect incorrect length" do
+      expect_raises(ULID::IncorrectLength) { ULID.validate!("0") }
+      expect_raises(ULID::IncorrectLength) { ULID.validate!("01B3EAF48P97R8MP9WS6MHDTZ32") }
+      expect_raises(ULID::IncorrectLength) { ULID.validate!("01B3EAF48P97R8MP9WS6MHDTZ") }
+      expect_raises(ULID::IncorrectLength) { ULID.validate!("!@#$%^&*(") }
+      expect_raises(ULID::IncorrectLength) { ULID.validate!("abcde") }
+      expect_raises(ULID::IncorrectLength) { ULID.validate!("1234567890") }
+      expect_raises(ULID::IncorrectLength) { ULID.validate!("") }
```
Comment (dukenguyenxyz) [inline spec/ulid_spec.cr]: just imo might be easier to write these specs with loop, i.e.

```crystal
["0", "abcde", ...].each { |i|
 expect_raises(ULID::IncorrectLength) {ULID.validate!(i)}
}
```

File: spec/ulid_spec.cr [hunk 25]
```diff
@@ -57,4 +57,82 @@ describe ULID do
...
+
+  describe ".seed_time" do
+    it "should correctly decode seed time" do
+      seedtime = ULID.seed_time("01B3EAF48P97R8MP9WS6MHDTZ3")
+      seedtime.should be_a Time
+      seedtime.should eq Time.unix_ms(1481170718998)
+
+      seedtime1 = ULID.seed_time("01EVDRF3VB5VD3211Z4DA112V9")
+      seedtime1.should be_a Time
+      seedtime1.should eq Time.unix_ms(1610000863083)
```
Comment (dukenguyenxyz) [inline spec/ulid_spec.cr]: imo dont need to compare type here since Time.unix will return a Time type, it should already handle comparing type on line 121

Comment (dukenguyenxyz) [reply spec/ulid_spec.cr]: im being too nitty-gritty now though

File: src/ulid.cr (hunk 23 above)
Comment (kimburgess) [reply src/ulid.cr]: @dukeraphaelng [`Nil` is more idiomatic for this case](https://crystal-lang.org/reference/syntax_and_semantics/return_types.html#nil-return-type).

@GabFitzgerald you can drop the `nil` on L45 though. As you've specified the method type as returning `Nil` it always provide that, regardless of what's in the method body.

File: src/ulid/exceptions.cr (hunk 20 above)
Comment (kimburgess) [reply src/ulid/exceptions.cr]: With the design of this you want to consider the how someone will use this lib. Calls to `.validate!` may occur somewhere deep within a call stack, with exception handling for this possible a little further away.

With the current approach a user must either create a [pokemon exception handler](http://wiki.c2.com/?PokemonExceptionHandling), which could capture any number of horrible things computers do, or individually catch both `ULID::IncorrectLength` _and_ `ULID::InvalidChars`. In reality all they want to know about is `ULID::Invalid`.

iIf you are modelling something with more diverse failure options a good pattern is a base exception for the domain that you are working in (`Foo::Error` or similar), and then more explicit error types that inherit from this. This was the base error can be caught for a general handler, or more specific ones where suitable too. In this case though I'd just have a single error type to keep things simple.

File: spec/ulid_spec.cr (hunk 24 above)
Comment (GabFitzgerald) [reply spec/ulid_spec.cr]: nice
